### ISSUE TITLE
How can I limit the total number of caches?

### ISSUE THREAD

**Issue by jpark58** [google/ExoPlayer#9208]
Hello,

I want to implement caching for a list of videos.
What I am trying to implement is pre-caching some chunk of every videos and store them in cache or disk so that users can watch videos without buffering.

I just want to limit the total number of caches. How do I limit the number? I want the size of video list to be the total number of caches so that LRU cache evictor can maintain only recent videos.

For example, 

limit of size of cache is 100kb for each file
```
DataSpec dataSpec = new DataSpec(videoUri, 0,1024*1024, null);
```

Then, limit of the number of caches...?

**Comment by marcbaechinger** [comment]
I think you have only one single `Cache` instance which is probably the provided `SimpleCache`. The you can use a `CacheEvictor` that makes sure the cache does not get too big.  This can be a `LeastRecentlyUsedCacheEvictor` but given your use case I would expect that you want another strategy because if the user skips forward only you don't want that the media that never has been played is evicted.  

I thinks it makes sense to have a working setup with an `LeastRecentlyUsedCacheEvictor` first. 

- create a single `SimpleCache` with a `LeastRecentlyUsedCacheEvictor` that you pass the max number of bytes to.
- Use `CachedDataSource.Factory` to create data sources for playback with the `SimpleCache` instance from above

This will limit the number of bytes the cache uses.

**Comment by jpark58** [comment]
Thank you!
I already have implemented the `LeastRecentlyUsedCacheEvictor` and of course set the max size of it. But, doesn't that mean it is literally the total size of cache files? I apologize if my wording above was misleading. **What exactly I am trying to implement is to limit the number of cache files.** 

Initializing variables for SimpleCache. By doing this, I believe I set 90MB as max size of the evictor.
```kotlin
private val exoPlayerCacheSize = 90 * 1024 * 1024L
private val leastRecentlyUsedCacheEvictor: LeastRecentlyUsedCacheEvictor = LeastRecentlyUsedCacheEvictor(exoPlayerCacheSize)
private val exoDatabaseProvider: ExoDatabaseProvider = ExoDatabaseProvider(context)
private val simpleCache = SimpleCache(dir, leastRecentlyUsedCacheEvictor, exoDatabaseProvider)
```

Here, I limited the size of each file to be cached and used it as a data spec in `CacheUtil.cache()` method.
```kotlin
val dataSpec = DataSpec(uri, 0, 1024 * 1024, null)
```

So basically, not only limiting the max size of total files and each file, but also I want to limit the number of files to be cached.

Thank you in advance

**Comment by marcbaechinger** [comment]
The `Cache` is not caching files directly but instead caches chunks of these files. ExoPlayer may only download parts of the files which can be seen eg. in the method of `CacheEvictor` that has a start and length for a give key that is representing a resource:

```
void onStartFile(Cache cache, String key, long position, long length);
```

I think I understand what you mean though. You want to only cache the begin of N streams/files. But this model is not how the `Cache` and the `CacheEvictor` internally works. You may be able to implement your own `CacheEvictor` that tries to track the keys that are passed to `onStartFile`. I don't think this is trivial to do though. I actually doubt whether this is worth it.

I honestly don't see the benefit of being able to limit the number of files. Because at least approximately you get this when you do `numberOfFiles * (1024 * 1024) = maxCacheSize` that you can pass to the `LeastRecentlyUsedCacheEvitor`. What is not sufficient with this solution? You users won't be concerned about having stored 11 or 87 files. They are concerned about having not enough space on the device which is independent from numbers of files.

**Comment by jpark58** [comment]
Thank you I really appreciate your help
